db.go

```diff
@@ -127,48 +154,119 @@ func (x *dummyAuthenticator) ResetPasswordFinish(identity string, token string,
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
+	if user, exists := x.users[userId]; exists {
+		user.email = email
+		user.username = username
+		user.firstname = firstname
+		user.lastname = lastname
+		user.mobilenumber = mobilenumber
+	} else {
+		return ErrIdentityAuthNotFound
+	}
+	return nil
+}
 
-	if _, exists := x.users[CanonicalizeIdentity(newIdent)]; exists {
-		return ErrIdentityExists
+func (x *dummyUserStoreAndAuth) ArchiveIdentity(userId UserId) error {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
+	if user, exists := x.users[userId]; exists {
+		user.archived = true
+	} else {
+		return ErrIdentityAuthNotFound
 	}
+	return nil
+}
 
-	oldIdent = CanonicalizeIdentity(oldIdent)
-	newKey := CanonicalizeIdentity(newIdent)
+func (x *dummyUserStoreAndAuth) RenameIdentity(oldEmail, newEmail string) error {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
 
-	if user, exists := x.users[oldIdent]; exists {
-		delete(x.users, oldIdent)
-		x.users[newKey] = &dummyUser{newIdent, user.password, ""}
-		return nil
+	newKey := CanonicalizeIdentity(newEmail)
+	oldEmail = CanonicalizeIdentity(oldEmail)
+	newUser := getUser(x.users, newKey)
+	if newUser == nil {
+		oldUser := getUser(x.users, oldEmail)
+
+		if oldUser != nil {
+			x.users[oldUser.userId].email = newEmail
+			return nil
+		} else {
+			return ErrIdentityAuthNotFound
+		}
 	} else {
-		return ErrIdentityAuthNotFound
+		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) GetIdentities() ([]string, error) {
+func (x *dummyUserStoreAndAuth) GetIdentities() ([]AuthUser, error) {
 	x.usersLock.RLock()
-	list := []string{}
-	for identity, _ := range x.users {
-		list = append(list, identity)
+	defer x.usersLock.RUnlock()
+
+	list := []AuthUser{}
+	for _, v := range x.users {
+		if v.archived == false {
+			list = append(list, AuthUser{v.userId, v.email, v.username, v.firstname, v.lastname, v.mobilenumber})
+		}
 	}
-	x.usersLock.RUnlock()
 	return list, nil
 }
 
-func (x *dummyAuthenticator) Close() {
+func (x *dummyUserStoreAndAuth) GetIdentityFromUserId(userId UserId) (string, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	for _, v := range x.users {
+		if v.userId == userId {
+			return v.email, nil
+		}
+	}
+
+	return "", ErrIdentityAuthNotFound
+}
+
+func (x *dummyUserStoreAndAuth) GetUserIdFromIdentity(identity string) (UserId, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	if CanonicalizeIdentity(identity) == "" {
+		return -1, ErrIdentityEmpty
+	}
+
+	for _, v := range x.users {
+		if v.email == identity {
+			return v.userId, nil
+		}
+	}
+	return -1, ErrIdentityAuthNotFound
+}
+
+func getUser(users map[UserId]*dummyUser, email string) *dummyUser {
+	for _, v := range users {
+		if CanonicalizeIdentity(v.email) == CanonicalizeIdentity(email) {
+			return v
+		}
+	}
+	return nil
+}
+
+func generateUserId(users map[UserId]*dummyUser) UserId {
+	return UserId(len(users) + 1)
```

Comment:
It would be far better here to mimick the technique used by real databases, which is to keep a `nextUserId` variable, and always return that value, then increment it by one, whenever creating a new user. I know this is only used for unit tests, but it's the right thing to do. Some day, somebody is going to waste an hour tracking down a strange unit test failure, when it would have been 1 minute with an incrementing counter. The technique used here is prone to generating the same ID twice, which is not something one would expect. The two different methods use almost the same number of code lines, so might as well use the more robust one.